### ISSUE TITLE
URI.to_string does not validate host

### ISSUE THREAD

**Issue by vans163** [elixir-lang/elixir#12188]
### Elixir and Erlang/OTP versions

Erlang/OTP 24 [erts-12.3.1] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit]

Elixir 1.13.1 (compiled with Erlang/OTP 24)

### Operating system

Ubuntu 22.04

### Current behavior

```elixir
URI.to_string %URI{scheme: "http", host: "bob.com///../../", path: "/hello"}
"http://bob.com///../..//hello"
```

But keys like `path` are validated

```elixir
URI.to_string %URI{scheme: "http", host: "bob.com", path: "hello"} 
** (ArgumentError) :path in URI must be empty or an absolute path if URL has a :host, got: %URI{scheme: "http", userinfo: nil, host: "bob.com", port: nil, path: "hello", query: nil, fragment: nil}
    (elixir 1.14.0) lib/uri.ex:981: String.Chars.URI.to_string/1
    iex:76: (file)

```

### Expected behavior

```elixir
URI.to_string %URI{scheme: "http", host: "bob.com///../../", path: "/hello"}
** (ArgumentError) :host in URI is invalid, got: %URI{scheme: "http", userinfo: nil, host: "bob.com///../../", port: nil, path: "/hello", query: nil, fragment: nil}
    (elixir 1.14.0) lib/uri.ex:981: String.Chars.URI.to_string/1
    iex:76: (file)
```

**Comment by josevalim** [comment]
It is not the job of `URI.to_string` to validate the URI, because doing so would make printing of URLs which have already been validated slow. Imagine if every URI operation always validated its contents.

We always assume it is valid and if there is a cheap validation we can perform, then we do it to catch obvious mistakes. If the goal is consistency, then we should rather remove the validation altogether.

**Comment by vans163** [comment]
> It is not the job of `URI.to_string` to validate the URI, because doing so would make printing of URLs which have already been validated slow. Imagine if every URI operation always validated its contents.
> 
> We always assume it is valid and if there is a cheap validation we can perform, then we do it to catch obvious mistakes. If the goal is consistency, then we should rather remove the validation altogether.

What about introducing a `URI.to_valid_string` I have noticed quite a bit of path and other validation logic included in 1.14 plus upstream erlang has a bit. It seems a little natural to extend this trend.

**Comment by josevalim** [comment]
I think the question is why is the URI invalid in the first place. We should rather not allow an invalid one to be created.
